internal/compare/compare_test.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compare
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/thriftrw/compile"
+)
+
+func TestErrorRequiredCase(t *testing.T) {
+	t.Parallel()
+	type test struct {
+		desc       string
+		fromStruct *compile.StructSpec
+		toStruct   *compile.StructSpec
+		wantError  string
+	}
+	tests := []test{
+		{
+			desc: "changed an optional field to required",
+			fromStruct: &compile.StructSpec{
+				Name: "structA",
+				Fields: compile.FieldGroup{
+					&compile.FieldSpec{
+						Required: false,
+						Name:     "fieldA",
+					},
+				},
+			},
+			toStruct: &compile.StructSpec{
+				Name: "structA",
+				Fields: compile.FieldGroup{
+					&compile.FieldSpec{
+						Required: true,
+						Name:     "fieldA",
+					},
+				},
+			},
+			wantError: "foo.thrift:changing an optional field fieldA in structA to required is not backwards compatible\n",
+		},
+		{
+			desc: "found a new required field",
+			fromStruct: &compile.StructSpec{
+				Fields: compile.FieldGroup{},
+			},
+			toStruct: &compile.StructSpec{
+				Name: "structA",
+				Fields: compile.FieldGroup{
+					&compile.FieldSpec{
+						Required: true,
+						Name:     "fieldA",
+					},
+				},
+			},
+			wantError: "foo.thrift:adding a required field fieldA to structA is not backwards compatible\n",
```

Comment:
nit: given that the program is for detecting backwards incompatible changes, we could skip that from the messages and just tell people what they're doing wrong:

    foo.thrift:adding a required field fieldA to structA